node/narwhal/src/gateway.rs

```diff
@@ -834,65 +834,132 @@ impl<N: Network> Gateway<N> {
+                    0u8..,
+                )
+                    .prop_map(|(a, b, c, d)| (a, b, c.account, Some(d as u16)))
+            })
+            .boxed()
+    }
+
+    fn any_valid_prod_gateway() -> BoxedStrategy<GatewayInput> {
+        (any::<CommitteeContext>(), any::<Selector>())
+            .prop_flat_map(|(context, account_selector)| {
+                let CommitteeContext(_, ValidatorSet(validators)) = context.clone();
+                (
+                    any_with::<Storage<CurrentNetwork>>(context.clone()),
+                    Just(context),
+                    Just(account_selector.select(validators)),
+                )
+                    .prop_map(|(a, b, c)| (a, b, c.account, None))
+            })
+            .boxed()
+    }
 
-        pub async fn generate_workers(
-            &self,
-            gateway: &Gateway<CurrentNetwork>,
-        ) -> (IndexMap<u8, Worker<CurrentNetwork>>, IndexMap<u8, WorkerSender<CurrentNetwork>>) {
+    #[proptest]
+    fn gateway_dev_initialization(#[strategy(any_valid_dev_gateway())] input: GatewayInput) {
+        let (storage, _, account, dev) = input;
+        let address = account.address();
+        let gateway = Gateway::new(account, storage, dev);
+        assert!(gateway.is_ok());
+        let gateway = gateway.unwrap();
+        let tcp_config = gateway.tcp().config();
+        assert_eq!(tcp_config.listener_ip, Some(IpAddr::V4(Ipv4Addr::LOCALHOST)));
+        assert_eq!(tcp_config.desired_listening_port, Some(MEMORY_POOL_PORT + dev.unwrap()));
+
+        let tcp_config = gateway.tcp().config();
+        assert_eq!(tcp_config.max_connections, MAX_COMMITTEE_SIZE);
+        assert_eq!(gateway.account().address(), address);
+    }
+
+    #[proptest]
+    fn gateway_prod_initialization(#[strategy(any_valid_prod_gateway())] input: GatewayInput) {
+        let (storage, _, account, dev) = input;
+        let address = account.address();
+        let gateway = Gateway::new(account, storage, dev);
+        assert!(gateway.is_ok());
+        assert!(dev.is_none());
+        let gateway = gateway.unwrap();
+        let tcp_config = gateway.tcp().config();
+        assert_eq!(tcp_config.listener_ip, Some(IpAddr::V4(Ipv4Addr::UNSPECIFIED)));
+        assert_eq!(tcp_config.desired_listening_port, Some(MEMORY_POOL_PORT));
+
+        let tcp_config = gateway.tcp().config();
+        assert_eq!(tcp_config.max_connections, MAX_COMMITTEE_SIZE);
+        assert_eq!(gateway.account().address(), address);
+    }
+
+    #[proptest(async = "tokio")]
+    async fn gateway_start(
+        #[strategy(any_valid_dev_gateway())] input: GatewayInput,
+        #[strategy(0..MAX_WORKERS)] workers_count: u8,
+    ) {
+        let (storage, _, account, dev) = input;
+        let worker_storage = storage.clone();
+        let gateway = Gateway::new(account, storage, dev);
+        assert!(gateway.is_ok());
+        let gateway = gateway.unwrap();
```

Comment:
```suggestion
        let gateway = gateway.unwrap();
```
(The test will also panic when this is an err.)